synapse/lib/filepath.py

```diff
@@ -0,0 +1,530 @@
+    '''
+    path = genpath(*paths)
+    if not os.path.exists(path):
+        raise s_exc.NoSuchPath(path=path)
+
+    if os.path.isdir(path):
+        return path_ctors.get('fs.reg.dir')
+    mime = _mimeFile(open(path, 'rb'))
+    return path_ctors.get(mime)
+
+def _fdClass(fd):
+    mime = _mimeFile(fd)
+    fd.seek(0)
+    return path_ctors.get(mime)
+
+def _mimeFile(fd):
+    '''
+    returns reg.file unless it is a known container
+    '''
+    try:
+        fd.seek(0)
+        tarfile.open(fileobj=fd)
+        return 'fs.tar.file'
+    except Exception as e:
+        pass
+
+    # OMG order matters here, zipfile will return True on some tar files
+    # check tar before zip
+    fd.seek(0)
+    if zipfile.is_zipfile(fd):
+        return 'fs.zip.file'
+
+    return 'fs.reg.file'
+
+def getPathParts(path):
+    '''
+    Returns the elements of a path in order, w/o regard to their original separators
+    '''
+
+    parts = path.replace('\\', '/').rstrip('/').split('/')
+
+    if parts[0] == '':
+        parts[0] = '/'
+
+    return parts
+
+def parsePaths(*paths):
+    '''
+    function to parse the incoming path.
+    lists of paths are joined prior to parsing
+    The path supports python's fnmatch glob matching
+
+    '''
+    if None in paths:
+        return None
+
+    path = genpath(*paths)
+
+    path_parts = getPathParts(path)
+
+    cls = _pathClass(path_parts[0])
+    base = cls(path_parts[0], parent=None)
+
+    bases = [base]
+    chld_bases = []
+    for name in path_parts[1:]:
+        chld_bases = []
+        for base in bases:
+            for member in base.list():
+                if not fnmatch.fnmatch(member, name):
+                    continue
+                chld_bases.append(base.child(member))
+
+        bases = chld_bases
+
+    if not bases:
+        return None
+
+    return bases
+
+def parsePath(*paths):
+    '''
+    function to parse the incoming path.
+    lists of paths are joined prior to parsing
+    '''
+    if None in paths:
+        return None
+
+    path = genpath(*paths)
+
+    path_parts = getPathParts(path)
+
+    try:
+
+        cls = _pathClass(path_parts[0])
+        base = cls(path_parts[0], parent=None)
+
+        for name in path_parts[1:]:
+            base = base.child(name)
+            if base == None:
+                return None
+
+    except s_exc.NoSuchPath as e:
+        return None
+
+    return base
+
+def openFiles(*paths, **kwargs):
+    '''
+    Yields a read-only file-like object for each path even if the path terminates inside a container file.
+    Paths may use python's fnmatch glob matching
+
+    If the path is a regular os accessible path mode may be passed through as a keyword argument.
+    If the path terminates in a container file, mode is ignored.
+
+    If the path does not exist a NoSuchPath exception is raised.
+    If req=True (Default) NoSuchPath will also be raised if ANY matching path exists, but is a directory
+
+    ex.
```

Comment:
For consistency with the other doc strings, maybe something like:

```
    Example:
        for fd in openFiles('/foo/bar/*.egg/dir0/zz*/nest.zip'):
            dostuff(fd)
```
I've always found that i'm better reminded about what a function does by seeing not only how it's called, but what is being done with a potential return value :D